src/Xamarin.Android.Build.Tasks/Tasks/DetermineJavaLibrariesToCompile.cs

```diff
@@ -61,6 +67,17 @@ public override bool Execute ()
 				foreach (var jar in AdditionalJavaLibraryReferences.Distinct (TaskItemComparer.DefaultComparer))
 					jars.Add (jar);
 
+			// In API 23, org.apache.http.* package was moved.
+			// See: https://developer.android.com/about/versions/marshmallow/android-6.0-changes#behavior-apache-http-client
+			if (int.TryParse (AndroidApiLevel, NumberStyles.Integer, CultureInfo.InvariantCulture, out int apiLevel) && apiLevel >= 23) {
```

Comment:
so we are always adding this jar file? Is that going to impact any project which dont need it?

Reply:
> so we are always adding this jar file?

For now, yes.

For example, the `org.apache.http.HttpRequest` type *was* in `android.jar`, and we have bound it as `Org.Apache.Http.IHttpRequest`, but [`org.apache.http.HttpRequest` was deprecated in API-22](https://developer.android.com/sdk/api_diff/22/changes/org.apache.http.HttpRequest) and has since been *removed* from `android.jar` and placed into `org.apache.http.legacy.jar` as of API-23.

Consequently, if an app uses the `Org.Apache.Http` namespace, the C# code will compile, but the JCW code might not, which would constitute an API break on our part.

We should absolutely do something "better," but said "better" fix hasn't been spec'd and would require more work than just always referencing `org.apache.http.legacy.jar` for now.

> Is that going to impact any project which dont need it?

Presumably, yes.

Short of inserting a task which scans all assembly references for these types, I'm not sure how we'd avoid that, short of a flag to say "Don't reference this library", which is inherently brittle.  (Set the flag, then add a NuGet which uses the type?  Your app now breaks.  Oops!)

Again, we *need* a better solution, but we're not sure what that better solution looks like yet.  (Probably a new `Org.Apache.Http.Legacy.dll` binding assembly, and type forwarders, and now we have a circular dependency between `Mono.Android.dll` and `Org.Apache.Http.Legacy.dll` (?!), and...)

Reply:
So this is only adding to `@(AndroidExternalJavaLibrary)`, which adds to:
* `--libraryjars` for ProGuard
* `--lib` for R8

This should allow these shrinkers *know* about these types and not give errors. But these types won't be included in the final `classes.dex`.

I think it will be OK for this to happen for all projects, since it shouldn't alter the final output.